contracts/src/BeefyClient.sol

```diff
@@ -390,49 +417,57 @@ contract BeefyClient {
-     *
-     * Research by W3F suggests that `ceil(log2(3 * x))` is a minimum number of signatures required to make an
-     * attack unfeasible. We put a further minimum bound of 10 on this value for extra security.
-     *
-     * If the session has less than 10 active validators it's definitely some sort of local testnet
-     * and we use different logic (minimum 2/3 + 1 validators must sign).
-     *
-     * One assumption is that Polkadot/Kusama will never have more than roughly 20,000 active validators in a session.
-     * As of writing this comment, Polkadot has 300 validators and Kusama has around 1000 validators,
-     * so we are well within those limits.
-     *
-     * In any case, an order of magnitude increase in validator set sizes will likely require a re-architecture
-     * of Polkadot that would make this contract obsolete well before the assumption becomes a problem.
-     *
-     * Constants generated with the help of scripts/minsigs.py
-     */
-    function minimumSignatureThreshold(uint256 validatorSetLen) internal pure returns (uint256) {
-        if (validatorSetLen <= 10) {
+    // Calculates the number of signature samples required by validator set length and the number of times a validator
+    // signature was used.
+    //
+    // ceil(log2(validatorSetLen)) + 1 * 2 ceil(log2(signatureUseCount))
+    //
+    // See https://hackmd.io/9OedC7icR5m-in_moUZ_WQ for full analysis.
+    function signatureSamples(uint256 validatorSetLen, uint256 signatureUseCount) internal view returns (uint256) {
+        // There are less validators than the minimum signatures so validate 2/3 majority.
+        if (validatorSetLen <= minimumSignatureSamples) {
             return validatorSetLen - (validatorSetLen - 1) / 3;
-        } else if (validatorSetLen < 342) {
-            return 10;
-        } else if (validatorSetLen < 683) {
-            return 11;
-        } else if (validatorSetLen < 1366) {
-            return 12;
-        } else if (validatorSetLen < 2731) {
-            return 13;
-        } else if (validatorSetLen < 5462) {
-            return 14;
-        } else if (validatorSetLen < 10923) {
-            return 15;
-        } else if (validatorSetLen < 21846) {
-            return 16;
-        } else {
-            return 17;
         }
+
+        // Start with the minimum number of signatures.
+        uint256 samples = minimumSignatureSamples;
+
+        // We must substrate minimumSignatures from the number of validators or we might end up
+        // requiring more signatures than there are validators.
+        uint256 extraValidators = validatorSetLen - minimumSignatureSamples;
+        // Using highest bit set to yield floor(log2(x))
+        uint256 validatorsLog2 = Bits.highestBitSet(extraValidators);
+        // We require ceil(log2(x)) so we round up if there were any extra bits set.
+        if (extraValidators > (1 << validatorsLog2)) {
+            validatorsLog2++;
+        }
```

Comment:
Lets factor this out into a `ceil` function in a `Math.sol` utility library

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/956/commits/2a9d732f2d84c0585ae3dee3ef191206b9eb6fb0.